apps/sms/views/conversation/js/conversation.js

```diff
@@ -1178,22 +1168,36 @@ var ConversationView = {
     return draftDiscardPromise.then(() => {
       // Now we'll try to find existing thread for the new message, otherwise
       // let's fallback to new message composer.
-      var threadPromise = Promise.reject();
       if (parameters && parameters.number) {
-        threadPromise = MessageManager.findThreadFromNumber(parameters.number);
+        // A rejected promise will be returned in case we can't find thread
+        // for the specified number.
+        return MessageManager.findThreadFromNumber(parameters.number);
       }
+      return Promise.reject();
+    }).then(
+      (id) => Navigation.toPanel('thread', { id: id, focusComposer: true }),
+      () => {
+        var draftPromise;
+        if (parameters.messageId) {
+          draftPromise = this.storeDraftFromMessage(parameters.messageId);
+        } else if (parameters.number) {
+          var draft = new Draft({
+            recipients: [parameters.number],
+            type: Utils.isEmailAddress(parameters.number) ? 'mms' : 'sms'
+          });
+          draftPromise = Drafts.add(draft).store().then(() => draft.id);
+        } else {
+          draftPromise = Promise.reject(new TypeError('Unknown parameter'));
+        }
 
-      // The rejected promise will be returned in case we can't find thread
-      // for the specified number.
-      return threadPromise.then((id) => {
-        return Navigation.toPanel('thread', { id: id, focusComposer: true });
-      }, () => {
-        return Navigation.toPanel('composer', {
-          activity: parameters,
-          focusComposer: !!(parameters && parameters.number)
-        });
-      });
-    });
+        return draftPromise.then(
+          (draftId) => Navigation.toPanel('composer', {
+            draftId,
+            focusComposer: !!(parameters && parameters.number)
+          })
+        );
+      }
+    );
```

Comment:
Mmmm, me too. I'll simplify this method a bit in my "remove-draft-dialog" PR,  but we can improve it in follow ups. 

__nit:__ Also maybe something like this will be easier to read, not ideal as well (as you noted below we may store parsed smil in DB so that we don't have SMIL dependency in Drafts)?
```
var draft;
if (parameters.messageId) {
  draft = Draft.fromMessage(Threads.Messages.get(messageId));
  // or something similar (e.g. ConversationView.draftFromMessage),
  // that does only one thing at time - creates Draft instance.
} else if (parameters.number) {
  draft = new Draft({
    recipients: [parameters.number],
    type: Utils.isEmailAddress(parameters.number) ? 'mms' : 'sms'
  });
} else {
  throw new TypeError('Unknown parameter');
}

return Drafts.add(draft).store().then(() =>
  Navigation.toPanel('composer', {
    draftId: draft.id,
    focusComposer: !!(parameters && parameters.number)
  })
);
```